### ISSUE TITLE
invalid operands to binary expression ('unsigned long' and 'Optional<uint64_t>' (aka 'Optional<unsigned long long>'))

### ISSUE THREAD

**Issue by ryandesign** [facebook/folly#602]
Hello, I'm a developer with the MacPorts project trying to update our folly port to version 2017.05.22.00, but it does not build for me on macOS Sierra 10.12.4 with Xcode 8.3.2 and clang 802.0.42 (Apple LLVM 8.1.0):

```
libtool: compile:  ccache /usr/bin/clang++ -DHAVE_CONFIG_H -I./.. -pthread -I/opt/local/include -I/opt/local/include -std=gnu++1y -pipe -Os -stdlib=libc++ -arch x86_64 -MT io/Compression.lo -MD -MP -M
F io/.deps/Compression.Tpo -c io/Compression.cpp  -fno-common -DPIC -o io/.libs/Compression.o
io/Compression.cpp:1458:24: error: invalid operands to binary expression ('unsigned long' and 'Optional<uint64_t>' (aka 'Optional<unsigned long long>'))
  if (decompressedSize != uncompressedLength) {
      ~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~
./../folly/Optional.h:378:6: note: candidate template ignored: deduced conflicting types for parameter 'V' ('unsigned long' vs. 'unsigned long long')
bool operator!=(const V& a, const Optional<V>& b) {
     ^
./../folly/Optional.h:368:6: note: candidate template ignored: could not match 'Optional<type-parameter-0-0>' against 'const unsigned long'
bool operator!=(const Optional<V>& a, const V& b) {
     ^
./../folly/Optional.h:390:6: note: candidate template ignored: could not match 'Optional<type-parameter-0-0>' against 'const unsigned long'
bool operator!=(const Optional<V>& a, const Optional<V>& b) {
     ^
./../folly/FBString.h:2558:6: note: candidate template ignored: could not match 'basic_fbstring<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2, type-parameter-0-3>' against 'const unsigned long'
bool operator!=(const basic_fbstring<E, T, A, S>& lhs,
     ^
./../folly/FBString.h:2564:6: note: candidate template ignored: could not match 'basic_fbstring' against 'Optional'
bool operator!=(const typename basic_fbstring<E, T, A, S>::value_type* lhs,
     ^
./../folly/FBString.h:2570:6: note: candidate template ignored: could not match 'basic_fbstring<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2, type-parameter-0-3>' against 'const unsigned long'
bool operator!=(const basic_fbstring<E, T, A, S>& lhs,
     ^
./../folly/FBString.h:2760:13: note: candidate template ignored: could not match 'basic_fbstring<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2, type-parameter-0-3>' against 'const unsigned long'
inline bool operator!=(
            ^
./../folly/FBString.h:2767:13: note: candidate template ignored: could not match 'basic_string<type-parameter-0-0, type-parameter-0-1, type-parameter-0-4>' against 'const unsigned long'
inline bool operator!=(
            ^
./../folly/Expected.h:747:13: note: candidate template ignored: could not match 'Unexpected<type-parameter-0-0>' against 'const unsigned long'
inline bool operator!=(
            ^
./../folly/Expected.h:1279:13: note: candidate template ignored: could not match 'Expected<type-parameter-0-0, type-parameter-0-1>' against 'const unsigned long'
inline bool operator!=(
            ^
./../folly/Expected.h:1367:6: note: candidate template ignored: could not match 'Expected<type-parameter-0-0, type-parameter-0-1>' against 'const unsigned long'
bool operator!=(const Expected<Value, Error>&, const Value& other) = delete;
     ^
./../folly/Expected.h:1379:6: note: candidate template ignored: could not match 'Expected' against 'Optional'
bool operator!=(const Value& other, const Expected<Value, Error>&) = delete;
     ^
1 error generated.
make[2]: *** [io/Compression.lo] Error 1
```

**Comment by lbrandy** [comment]
cc: @terrelln 

I actually think this is a bug in https://github.com/facebook/folly/blob/master/folly/Optional.h#L362. It's forcing the types to be the same in `operator!=` even if they are convertible to one another. In this case we're trying to use `unsigned long` and `unsigned long long`.

**Comment by lbrandy** [comment]
I put up a small fix internally for review -> D5112464.
